File: packages/backend/src/app/api/ng_list_domains.py
"""
NGリストドメイン管理APIルーター

NGリストドメインのCRUD操作とチェック機能のエンドポイントを提供します。
"""

from typing import Annotated

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from src.app.api.dependencies import get_current_active_user
from src.app.core.database import get_db
from src.application.schemas.ng_list_domain import (
    NgListDomainCheckRequest,
    NgListDomainCheckResponse,
    NgListDomainCreateRequest,
    NgListDomainListResponse,
    NgListDomainResponse,
)
from src.application.use_cases.ng_list_domain_use_cases import NgListDomainUseCases
from src.domain.entities.ng_list_domain_entity import NgListDomainEntity
from src.domain.entities.user_entity import UserEntity
from src.domain.interfaces.list_repository import IListRepository
from src.domain.interfaces.ng_list_domain_repository import INgListDomainRepository
from src.infrastructure.persistence.repositories.list_repository import ListRepository
from src.infrastructure.persistence.repositories.ng_list_domain_repository import (
    NgListDomainRepository,
)

router = APIRouter(prefix="/ng-list-domains", tags=["ng-list-domains"])


async def get_ng_list_domain_use_cases(
    session: AsyncSession = Depends(get_db),
) -> NgListDomainUseCases:
    """
    NGリストドメインユースケースの依存性注入

    Args:
        session: データベースセッション

    Returns:
        NgListDomainUseCases: NGリストドメインユースケースインスタンス
    """
    ng_list_domain_repo: INgListDomainRepository = NgListDomainRepository(session)
    list_repo: IListRepository = ListRepository(session)
    return NgListDomainUseCases(
        ng_list_domain_repository=ng_list_domain_repo,
        list_repository=list_repo,
    )


@router.post(
    "",
    response_model=NgListDomainResponse,
    status_code=status.HTTP_201_CREATED,
    summary="NGドメイン追加",
    description="リストにNGドメインを追加します。認証が必要です。",
)
async def create_ng_domain(
    request: NgListDomainCreateRequest,
    current_user: UserEntity = Depends(get_current_active_user),
    use_cases: NgListDomainUseCases = Depends(get_ng_list_domain_use_cases),
) -> NgListDomainResponse:
    """
    NGドメインを追加

    - **list_id**: リストID
    - **domain**: NGドメインパターン（例: "example.com", "*.example.com"）
    - **memo**: メモ（任意）
    """
    ng_domain_entity = await use_cases.add_ng_domain(
        requesting_organization_id=current_user.organization_id,
        request=request,
    )
    return _to_response(ng_domain_entity)


@router.get(
    "/{ng_domain_id}",
    response_model=NgListDomainResponse,
    summary="NGドメイン取得",
    description="IDでNGドメインを取得します。認証が必要です。",
)
async def get_ng_domain(
    ng_domain_id: int,
    current_user: UserEntity = Depends(get_current_active_user),
    use_cases: NgListDomainUseCases = Depends(get_ng_list_domain_use_cases),
) -> NgListDomainResponse:
    """
    NGドメインを取得

    - **ng_domain_id**: NGドメインID
    """
    ng_domain_entity = await use_cases.get_ng_domain(
        ng_domain_id=ng_domain_id,
        requesting_organization_id=current_user.organization_id,
    )
    return _to_response(ng_domain_entity)


@router.get(
    "",
    response_model=NgListDomainListResponse,
    summary="NGドメイン一覧取得",
    description="リストに属するNGドメインの一覧を取得します。認証が必要です。",
)
async def list_ng_domains(
    list_id: Annotated[int, Query(gt=0, description="リストID")],
    current_user: UserEntity = Depends(get_current_active_user),
    use_cases: NgListDomainUseCases = Depends(get_ng_list_domain_use_cases),
) -> NgListDomainListResponse:
    """
    リストに属するNGドメインの一覧を取得

    - **list_id**: リストID
    """
    ng_domains = await use_cases.list_ng_domains_by_list(
        list_id=list_id,
        requesting_organization_id=current_user.organization_id,
    )

    return NgListDomainListResponse(
        ng_domains=[_to_response(ng_domain) for ng_domain in ng_domains],
        total=len(ng_domains),
    )


@router.delete(
    "/{ng_domain_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="NGドメイン削除",
    description="NGドメインを論理削除します。認証が必要です。",
)
async def delete_ng_domain(
    ng_domain_id: int,
    current_user: UserEntity = Depends(get_current_active_user),
    use_cases: NgListDomainUseCases = Depends(get_ng_list_domain_use_cases),
) -> None:
    """
    NGドメインを論理削除

    - **ng_domain_id**: NGドメインID
    """
    await use_cases.delete_ng_domain(
        ng_domain_id=ng_domain_id,
        requesting_organization_id=current_user.organization_id,
    )


@router.post(
    "/check",
    response_model=NgListDomainCheckResponse,
    summary="URLのNGチェック",
    description="URLがNGリストに含まれるかチェックします。認証が必要です。",
)
async def check_url_is_ng(
    request: NgListDomainCheckRequest,
    current_user: UserEntity = Depends(get_current_active_user),
    use_cases: NgListDomainUseCases = Depends(get_ng_list_domain_use_cases),
) -> NgListDomainCheckResponse:
    """
    URLがNGリストに含まれるかチェック

    - **list_id**: リストID
    - **url**: チェック対象のURL（例: "https://www.example.com/form"）

    Returns:
        - **is_ng**: NGリストに含まれる場合True
        - **matched_pattern**: マッチしたNGドメインパターン（NGの場合のみ）
        - **extracted_domain**: URLから抽出されたドメイン
    """
    is_ng, matched_pattern, extracted_domain = await use_cases.check_url_is_ng(
        list_id=request.list_id,
        url=request.url,
        requesting_organization_id=current_user.organization_id,
    )

    return NgListDomainCheckResponse(
        is_ng=is_ng,
        matched_pattern=matched_pattern,
        extracted_domain=extracted_domain,
    )


def _to_response(ng_domain_entity: NgListDomainEntity) -> NgListDomainResponse:
    """
    NgListDomainEntityをNgListDomainResponseに変換

    Args:
        ng_domain_entity: NGドメインエンティティ

    Returns:
        NgListDomainResponse: NGドメインレスポンス
    """
    return NgListDomainResponse(
        id=ng_domain_entity.id,
        list_id=ng_domain_entity.list_id,
        domain=ng_domain_entity.domain,
        domain_pattern=ng_domain_entity.domain_pattern,
        is_wildcard=ng_domain_entity.is_wildcard,
        memo=ng_domain_entity.memo,
        created_at=ng_domain_entity.created_at,
        updated_at=ng_domain_entity.updated_at,
        deleted_at=ng_domain_entity.deleted_at,
    )

File: packages/frontend/lib/api/ngListDomains.ts
/**
 * NGリストドメインAPI クライアント関数
 */

import { z } from 'zod'
import { get, post, del } from '../api-client'
import type {
  NgListDomain,
  NgListDomainListResponse,
  NgListDomainCreateRequest,
  NgListDomainCheckRequest,
  NgListDomainCheckResponse,
} from '@/types/ngListDomain'

/**
 * APIレスポンススキーマ（バックエンドのsnake_case形式）
 */
const ngListDomainResponseSchema = z.object({
  id: z.number(),
  list_id: z.number(),
  domain: z.string(),
  domain_pattern: z.string(),
  is_wildcard: z.boolean(),
  memo: z.string().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
  deleted_at: z.string().nullable().optional(),
})

/**
 * APIレスポンスをフロントエンドの型に変換
 * バックエンドのsnake_caseをcamelCaseに変換
 *
 * @param data - 未検証のAPIレスポンスデータ
 * @returns 検証済みのNgListDomainオブジェクト
 * @throws ZodError - レスポンスが期待する型と一致しない場合
 */
function transformNgListDomain(data: unknown): NgListDomain {
  // ランタイム型検証
  const validated = ngListDomainResponseSchema.parse(data)

  return {
    id: validated.id,
    listId: validated.list_id,
    domain: validated.domain,
    domainPattern: validated.domain_pattern,
    isWildcard: validated.is_wildcard,
    memo: validated.memo ?? null,
    createdAt: validated.created_at,
    updatedAt: validated.updated_at,
    deletedAt: validated.deleted_at ?? null,
  }
}

/**
 * リストレスポンススキーマ
 */
const ngListDomainListResponseSchema = z.object({
  ng_domains: z.array(ngListDomainResponseSchema),
  total: z.number(),
})

/**
 * リストに属するNGドメイン一覧を取得
 *
 * @param listId - リストID
 * @returns NGドメイン一覧
 */
export async function fetchNgListDomains(
  listId: number
): Promise<NgListDomainListResponse> {
  const response = await get<unknown>(
    `/ng-list-domains?list_id=${listId}`
  )

  // ランタイム型検証
  const validated = ngListDomainListResponseSchema.parse(response)

  return {
    ngDomains: validated.ng_domains.map(transformNgListDomain),
    total: validated.total,
  }
}

/**
 * NGドメインを取得
 *
 * @param ngDomainId - NGドメインID
 * @returns NGドメイン
 */
export async function fetchNgListDomain(ngDomainId: number): Promise<NgListDomain> {
  const response = await get<unknown>(`/ng-list-domains/${ngDomainId}`)
  return transformNgListDomain(response)
}

/**
 * NGドメインを追加
 *
 * @param data - 作成データ
 * @returns 作成されたNGドメイン
 */
export async function createNgListDomain(
  data: NgListDomainCreateRequest
): Promise<NgListDomain> {
  // camelCaseをsnake_caseに変換
  const requestData = {
    list_id: data.listId,
    domain: data.domain,
    memo: data.memo,
  }

  const response = await post<unknown>('/ng-list-domains', requestData)
  return transformNgListDomain(response)
}

/**
 * NGドメインを削除
 *
 * @param ngDomainId - NGドメインID
 */
export async function deleteNgListDomain(ngDomainId: number): Promise<void> {
  await del(`/ng-list-domains/${ngDomainId}`)
}

/**
 * チェックレスポンススキーマ
 */
const ngListDomainCheckResponseSchema = z.object({
  is_ng: z.boolean(),
  matched_pattern: z.string().nullable().optional(),
  extracted_domain: z.string().nullable().optional(),
})

/**
 * URLがNGリストに含まれるかチェック
 *
 * @param data - チェックリクエスト
 * @returns チェック結果
 */
export async function checkUrlIsNg(
  data: NgListDomainCheckRequest
): Promise<NgListDomainCheckResponse> {
  // camelCaseをsnake_caseに変換
  const requestData = {
    list_id: data.listId,
    url: data.url,
  }

  const response = await post<unknown>('/ng-list-domains/check', requestData)

  // ランタイム型検証
  const validated = ngListDomainCheckResponseSchema.parse(response)

  return {
    isNg: validated.is_ng,
    matchedPattern: validated.matched_pattern ?? null,
    extractedDomain: validated.extracted_domain ?? null,
  }
}
